src/solc_0.8/polygon/child/asset/PolygonAssetV2.sol

```diff
@@ -0,0 +1,86 @@
+        uint8 chainIndex
+    ) external {
+        require(trustedForwarder.isContract(), "TRUSTERFORWARDER_NOT_CONTRACT");
+        require(predicate.isContract(), "PREDICATE_NOT_CONTRACT");
+        require(childChainManager.isContract(), "CHILDCHAINMANAGER_NOT_CONTRACT");
+        initV2(trustedForwarder, admin, bouncerAdmin, predicate, chainIndex);
+        _childChainManager = childChainManager;
+    }
+
+    /// @notice called when tokens are deposited on root chain
+    /// @dev Should be callable only by ChildChainManager
+    /// @dev Should handle deposit by minting the required tokens for user
+    /// @dev Make sure minting is done only by this function
+    /// @param user user address for whom deposit is being done
+    /// @param depositData abi encoded ids array and amounts array
+    function deposit(address user, bytes calldata depositData) external onlyDepositor {
+        require(user != address(0), "INVALID_DEPOSIT_USER");
+        (uint256[] memory ids, uint256[] memory amounts, bytes memory data) =
+            abi.decode(depositData, (uint256[], uint256[], bytes));
+        address sender = _msgSender();
+        for (uint256 i = 0; i < ids.length; i++) {
+            // ERC-721
+            if ((amounts[i] == 1) && (ids[i] & IS_NFT > 0)) {
+                // temp: to be replaced by encoded IPFS hash
+                bytes32 dummyHash = bytes32("0x00");
+                uint8 rarity = 0;
+                _mint(dummyHash, amounts[i], rarity, sender, user, ids[0], data, false);
+            }
+            // ERC-1155
+            else {
+                // temp: to be replaced by encoded IPFS hash
+                bytes32 dummyHash = bytes32("0x00");
+                uint8 rarity = 0;
+                _mint(dummyHash, amounts[i], rarity, sender, user, ids[0], data, false);
+            }
+        }
+    }
+
+    /// @notice called when user wants to withdraw single token back to root chain
+    /// @dev Should burn user's tokens. This transaction will be verified when exiting on root chain
+    /// @param id id to withdraw
+    /// @param amount amount to withdraw
+    function withdrawSingle(uint256 id, uint256 amount) external {
+        _burn(_msgSender(), id, amount);
+        emit ChainExit(_msgSender(), id, amount, "");
+    }
+
+    /// @notice called when user wants to batch withdraw tokens back to root chain
+    /// @dev Should burn user's tokens. This transaction will be verified when exiting on root chain
+    /// @param ids ids to withdraw
+    /// @param amounts amounts to withdraw
+    function withdrawBatch(uint256[] calldata ids, uint256[] calldata amounts) external {
+        _burnBatch(_msgSender(), ids, amounts);
+        emit ChainExitBatch(_msgSender(), ids, amounts, "");
+    }
+
+    /// @notice Throws if called by any address other than depositor
+    modifier onlyDepositor() {
+        require(_msgSender() == _childChainManager, "!DEPOSITOR");
+        _;
+    }
```

Comment:
Minor, but if contract size is still an issue, this modifier which is only used once could probably be removed and just add the `require` statement directly to the `deposit` function.

Reply:
Yeah that would help with the size, thanks :+1: